sdk/nodejs/serverless/function.ts

```diff
@@ -0,0 +1,184 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import * as pulumi from "@pulumi/pulumi";
+import * as cloudfunctions from "../cloudfunctions";
+import * as storage from "../storage";
+
+import * as filepath from "path";
+import * as readPackageJson from "read-package-json";
+
+/**
+ * Handler is the signature for a serverless function.
+ */
+export type Handler = (request: any, response: any) => any;
+
+/**
+ * FunctionOptions provides configuration options for the serverless Function.
+ */
+export interface FunctionOptions {
+    /**
+     * Memory (in MB), available to the function. Default value is 256MB. Allowed values are: 128MB, 256MB, 512MB, 1024MB, and 2048MB.
+     */
+    readonly availableMemoryMb?: pulumi.Input<number>;
+    /**
+     * Description of the function.
+     */
+    readonly description?: pulumi.Input<string>;
+    /**
+     * A set of key/value label pairs to assign to the function.
+     */
+    readonly labels?: pulumi.Input<{
+        [key: string]: any;
+    }>;
+    /**
+     * Project of the function. If it is not provided, the provider project is used.
+     */
+    readonly project?: pulumi.Input<string>;
+    /**
+     * Region of function. Currently can be only "us-central1". If it is not provided, the provider region is used.
+     */
+    readonly region?: pulumi.Input<string>;
+    /**
+     * Timeout (in seconds) for the function. Default value is 60 seconds. Cannot be more than 540 seconds.
+     */
+    readonly timeout?: pulumi.Input<number>;        
+   /**
+    * The packages relative to the program folder to not include the Function upload. This can be
+    * used to override the default serialization logic that includes all packages referenced by
+    * project.json (except @pulumi packages).  Default is `[]`.
+    */
+   excludePackages?: string[];
+}
```

Comment:
We made several mistakes with `aws.serverless.Function` that ideally we wouldn't replicate here (and plan to fix in AWS).  Most notably, we don't really want to re-invent the set of options for a function here - we'd prefer to accept a `gcp.cloudfunctions.FunctionArgs` as input and merge it's options in directly with the source/entrypoint generated from the callback.  (Likely failing if conflicting settings are provided in the function args).

/cc @CyrusNajmabadi who is thinking about this for AWS and Azure as well.

Reply:
It's to you ofc, but sounds a bit leaky to me. We expose properties that can't be set (not just entrypoint, but e.g. `triggerHttp`). And we need an additional parameter `excludePackages` (AWS has more).
Serverless and raw are on different abstraction layers, so I'm not sure that options should be the same...

Reply:
> It's to you ofc, but sounds a bit leaky to me.

'leaky' is one way to put it :)   But 'opinionated' is how i'd like ot put it.  Specifically, these APIs intentionally exist to create the real cloud resource.  i.e. in AWS we are trying to exactly expose the ability to create an AWS lambda.  In Azure we we are trying to exactly expose the abilty to create an Azure FunctionApp.  And the same will be true for GCP (i'm not knowledgable enough there to know precisely what that is).

Because we are helping you create those exact Resources, we want to give you the exact same level of control.  The purpose of these APIs is not really to abstract that away.  But simply to do the heavy lifting of allowing you to provide a javascript object that is then converted to the Cloud Lambda/FunctionApp/etc. iimplementation.

Reply:
Do we still want `excludePackages` property? Do something like
``` typescript
export interface FunctionOptions extends cloudfunctions.FunctionArgs {   
   excludePackages?: string[];
}
```
or do we want to accept exactly `FunctionArgs`?

Reply:
@mikhailshilkov I think it's reasonable to do that (and, in fact, we should move closer to that in aws-serverless).  We do want to allow users to pass all (or nearly all) of the options to configure the function.  On top of that, we would want things like excludePackages to control the parts of logic that Pulumi adds itself (like packing up the code).  So i would go with this latest suggestion so that you're not blocked here.  We may end up tweaking this a bit as we solidify aws-serverless and whatnot.

Reply:
Hmm... This turns out problematic because the original `FunctionArgs` have parameters like `sourceArchiveBucket` mandatory, while `FunctionOptions` should actually depend *not* to have them... I guess I can't declare a type based on the other type, but *excluding* given properties?

Reply:
You *technically* can, using some advanced typescript features.  You can see one of the ytpes that can handle that here: https://github.com/pulumi/pulumi-aws-serverless/pull/31/files#diff-7d9b0842656b84e670d724366d103be7

Effectively, you could say: 

```ts
type FunctionArgs = Omit<cloudfunctions.FunctionArgs, "sourceArchiveBucket" | "otherPropName" | "etc">
```

However, i'm hesitant about thsi approach because it can be hard to comprehend.  i.e. it helps TS strongly check that your types have hte right shape.  But it can then be harder for someone to grok waht is actually going on.